File: apps/ios/CoreHub/Phone/DictationLanguage.swift
// Which language the phone's recognizer listens in (owner, 2026-09-27: «المفروض هو يتعرف
// تلقائي»). The iPhone's recognizer takes one locale, so «Auto» — the default — picks it the way a
// person would expect: the keyboard they are typing with, else the language the conversation is
// written in, else the phone's own languages, else the app's. Any language the phone can
// recognize can also be chosen (the long press on the microphone); none is special.
//
// The rules are here, apart from UIKit and Speech, so they are tested.
import Foundation

enum DictationLanguage {
    /// The stored choice that means "pick it for me".
    static let auto = "auto"

    /// Offered after the keyboard's languages in the microphone's menu: widely spoken ones.
    static let popular = ["en", "ar", "es", "fr", "de", "zh", "hi", "pt", "ja", "ru", "it", "tr", "ko", "id", "ur", "fa"]

    /// A writing system the conversation can be recognized by, cheaply, from its letters.
    enum Script: Equatable {
        case arabic, latin, cyrillic, devanagari, han, kana, hangul, hebrew, greek, thai
    }

    /// A keyboard's `primaryLanguage` (or any locale identifier) as a BCP-47 tag; nil for a
    /// keyboard that is not a language (emoji, dictation, handwriting).
    static func tag(_ raw: String?) -> String? {
        guard let trimmed = raw?.trimmingCharacters(in: .whitespacesAndNewlines), !trimmed.isEmpty else { return nil }
        var parts = trimmed.replacingOccurrences(of: "_", with: "-").split(separator: "-").map(String.init)
        guard let first = parts.first else { return nil }
        let base = first.lowercased()
        guard (2...3).contains(base.count), base.allSatisfy({ $0.isASCII && $0.isLetter }) else { return nil }
        if ["mul", "und", "zxx"].contains(base) { return nil }
        parts[0] = base
        // Chinese keyboards name the script; the recognizer names the region.
        if base == "zh", parts.count > 1 {
            switch parts[1].lowercased() {
            case "hans": return "zh-CN"
            case "hant": return "zh-TW"
            default: break
            }
        }
        if parts.count > 1 {
            let second = parts[1]
            parts[1] = second.count == 2 ? second.uppercased() : second.prefix(1).uppercased() + second.dropFirst().lowercased()
        }
        return parts.joined(separator: "-")
    }

    /// The language part of a tag: `ar` of `ar-SA`.
    static func base(_ tag: String) -> String {
        String(tag.replacingOccurrences(of: "_", with: "-").split(separator: "-").first ?? "").lowercased()
    }

    /// The script most letters of `texts` are written in; nil with fewer than three letters.
    static func script(of texts: [String]) -> Script? {
        var counts: [Script: Int] = [:]
        for text in texts {
            for scalar in text.unicodeScalars {
                guard let found = script(of: scalar) else { continue }
                counts[found, default: 0] += 1
            }
        }
        // Japanese mixes kana with Han: any kana makes it Japanese.
        if let kana = counts[.kana], kana > 0 {
            counts[.kana] = kana + (counts[.han] ?? 0)
            counts[.han] = nil
        }
        guard let best = counts.max(by: { $0.value < $1.value }), best.value >= 3 else { return nil }
        return best.key
    }

    private static func script(of scalar: Unicode.Scalar) -> Script? {
        switch scalar.value {
        case 0x0041...0x005A, 0x0061...0x007A, 0x00C0...0x024F, 0x1E00...0x1EFF: return .latin
        case 0x0600...0x06FF, 0x0750...0x077F, 0x08A0...0x08FF, 0xFB50...0xFDFF, 0xFE70...0xFEFF: return .arabic
        case 0x0400...0x052F: return .cyrillic
        case 0x0900...0x097F: return .devanagari
        case 0x3040...0x30FF: return .kana
        case 0x4E00...0x9FFF, 0x3400...0x4DBF: return .han
        case 0xAC00...0xD7AF, 0x1100...0x11FF: return .hangul
        case 0x0590...0x05FF: return .hebrew
        case 0x0370...0x03FF: return .greek
        case 0x0E00...0x0E7F: return .thai
        default: return nil
        }
    }

    /// The script a language is written in (the common case; Latin for the rest).
    static func script(ofLanguage tag: String) -> Script {
        switch base(tag) {
        case "ar", "fa", "ur", "ps", "ckb", "sd", "ug": return .arabic
        case "ru", "uk", "be", "bg", "sr", "mk", "kk", "ky", "mn", "tg": return .cyrillic
        case "hi", "mr", "ne", "sa": return .devanagari
        case "zh", "yue": return .han
        case "ja": return .kana
        case "ko": return .hangul
        case "he", "iw", "yi": return .hebrew
        case "el": return .greek
        case "th": return .thai
        default: return .latin
        }
    }

    /// The language written in `script` among `known` (the keyboards, then the phone's own
    /// languages), else the script's most spoken one.
    static func language(for script: Script, known: [String]) -> String {
        if let mine = known.compactMap(tag).first(where: { self.script(ofLanguage: $0) == script || (script == .han && base($0) == "ja") }) {
            // A Japanese reader's Han-only text is still Japanese.
            return mine
        }
        switch script {
        case .arabic: return "ar"
        case .latin: return "en"
        case .cyrillic: return "ru"
        case .devanagari: return "hi"
        case .han: return "zh"
        case .kana: return "ja"
        case .hangul: return "ko"
        case .hebrew: return "he"
        case .greek: return "el"
        case .thai: return "th"
        }
    }

    /// The languages to try, best first: the person's choice; with Auto, the keyboard in use,
    /// then the language the conversation is written in, then the phone's languages, then the
    /// app's. Duplicates are dropped.
    static func candidates(
        choice: String,
        keyboard: String?,
        recent: [String],
        keyboards: [String],
        preferred: [String],
        app: String
    ) -> [String] {
        var list: [String] = []
        if choice != auto, let chosen = tag(choice) { list.append(chosen) }
        if let keyboard = tag(keyboard) { list.append(keyboard) }
        if let script = script(of: recent) {
            list.append(language(for: script, known: keyboards + preferred))
        }
        list += preferred.compactMap(tag)
        if let app = tag(app) { list.append(app) }
        var seen = Set<String>()
        return list.filter { seen.insert($0.lowercased()).inserted }
    }

    /// The region a bare language is recognized in when the recognizer offers several.
    private static let usualRegion: [String: String] = [
        "ar": "SA", "en": "US", "es": "ES", "fr": "FR", "de": "DE", "pt": "BR", "zh": "CN",
        "it": "IT", "nl": "NL", "ru": "RU", "ja": "JP", "ko": "KR", "hi": "IN", "tr": "TR",
    ]

    /// The first candidate the recognizer supports, as one of `supported`'s own identifiers:
    /// the exact locale when it is there, else the same language in its usual region, else
    /// that language in any region.
    static func match(_ candidates: [String], supported: [String]) -> String? {
        let known = supported.map { (id: $0, tag: tag($0) ?? $0) }
        for candidate in candidates {
            guard let wanted = tag(candidate) else { continue }
            if let exact = known.first(where: { $0.tag.caseInsensitiveCompare(wanted) == .orderedSame }) { return exact.id }
            let language = base(wanted)
            let same = known.filter { base($0.tag) == language }.sorted { $0.tag < $1.tag }
            guard !same.isEmpty else { continue }
            if let region = usualRegion[language],
               let usual = same.first(where: { $0.tag.caseInsensitiveCompare("\(language)-\(region)") == .orderedSame }) {
                return usual.id
            }
            return same[0].id
        }
        return nil
    }

    /// What the microphone's menu offers, in order: the keyboards' languages, then the popular
    /// ones, each once and only when the phone can recognize it (`supported` nil: all).
    static func menu(keyboards: [String], supported: [String]?) -> [String] {
        var seen = Set<String>()
        var list: [String] = []
        for language in keyboards.compactMap(tag) + popular {
            let key = base(language)
            guard seen.insert(key).inserted else { continue }
            if let supported, match([language], supported: supported) == nil { continue }
            list.append(language)
        }
        return list
    }

    /// The tiny mark on the microphone while a language is chosen: `AR`, `EN`, `ZH`.
    static func badge(_ choice: String) -> String? {
        choice == auto ? nil : base(choice).uppercased()
    }

    /// A language's name in the reader's language: «العربية», "Arabic".
    static func name(_ tag: String, in language: String) -> String {
        let reader = Locale(identifier: language)
        let name = reader.localizedString(forIdentifier: tag) ?? reader.localizedString(forLanguageCode: base(tag)) ?? tag
        return name.prefix(1).uppercased() + name.dropFirst()
    }
}

File: apps/android/app/src/main/java/hub/core/android/phone/DictationLanguage.kt
package hub.core.android.phone

import java.util.Locale

/*
 * Which language the phone's recognizer listens in (owner, 2026-09-27: «المفروض هو يتعرف
 * تلقائي»). «Auto» — the default — follows the keyboard the person types with (its current
 * subtype), else the language the conversation is written in, else the phone's own languages,
 * else the app's. Any language the phone can recognize can be chosen from the microphone's long
 * press; none is special. The rules are here, apart from Android, so they are tested.
 */
object DictationLanguage {
    /** The stored choice that means "pick it for me". */
    const val AUTO = "auto"

    /** Offered after the keyboards' languages in the microphone's menu: widely spoken ones. */
    val popular = listOf("en", "ar", "es", "fr", "de", "zh", "hi", "pt", "ja", "ru", "it", "tr", "ko", "id", "ur", "fa")

    /** A writing system the conversation can be recognized by, cheaply, from its letters. */
    enum class Script { ARABIC, LATIN, CYRILLIC, DEVANAGARI, HAN, KANA, HANGUL, HEBREW, GREEK, THAI }

    /** A keyboard subtype's language (or any locale id) as a BCP-47 tag; null for none (emoji, handwriting). */
    fun tag(raw: String?): String? {
        val trimmed = raw?.trim()?.takeIf { it.isNotEmpty() } ?: return null
        val parts = trimmed.replace('_', '-').split('-').filter { it.isNotEmpty() }.toMutableList()
        val base = parts.firstOrNull()?.lowercase() ?: return null
        if (base.length !in 2..3 || !base.all { it in 'a'..'z' } || base in setOf("mul", "und", "zxx")) return null
        parts[0] = base
        if (base == "zh" && parts.size > 1) {
            when (parts[1].lowercase()) {
                "hans" -> return "zh-CN"
                "hant" -> return "zh-TW"
            }
        }
        if (parts.size > 1) {
            val second = parts[1]
            parts[1] = if (second.length == 2) second.uppercase() else second.lowercase().replaceFirstChar { it.uppercase() }
        }
        return parts.joinToString("-")
    }

    /** The language part of a tag: `ar` of `ar-SA`. */
    fun base(tag: String): String = tag.replace('_', '-').substringBefore('-').lowercase()

    /** The script most letters of [texts] are written in; null with fewer than three letters. */
    fun script(texts: List<String>): Script? {
        val counts = mutableMapOf<Script, Int>()
        texts.forEach { text ->
            var i = 0
            while (i < text.length) {
                val point = text.codePointAt(i)
                scriptOf(point)?.let { counts[it] = (counts[it] ?: 0) + 1 }
                i += Character.charCount(point)
            }
        }
        // Japanese mixes kana with Han: any kana makes it Japanese.
        counts[Script.KANA]?.let { kana ->
            counts[Script.KANA] = kana + (counts.remove(Script.HAN) ?: 0)
        }
        val best = counts.maxByOrNull { it.value } ?: return null
        return if (best.value >= 3) best.key else null
    }

    private fun scriptOf(point: Int): Script? = when (point) {
        in 0x41..0x5A, in 0x61..0x7A, in 0xC0..0x24F, in 0x1E00..0x1EFF -> Script.LATIN
        in 0x600..0x6FF, in 0x750..0x77F, in 0x8A0..0x8FF, in 0xFB50..0xFDFF, in 0xFE70..0xFEFF -> Script.ARABIC
        in 0x400..0x52F -> Script.CYRILLIC
        in 0x900..0x97F -> Script.DEVANAGARI
        in 0x3040..0x30FF -> Script.KANA
        in 0x4E00..0x9FFF, in 0x3400..0x4DBF -> Script.HAN
        in 0xAC00..0xD7AF, in 0x1100..0x11FF -> Script.HANGUL
        in 0x590..0x5FF -> Script.HEBREW
        in 0x370..0x3FF -> Script.GREEK
        in 0xE00..0xE7F -> Script.THAI
        else -> null
    }

    /** The script a language is written in (the common case; Latin for the rest). */
    fun scriptOfLanguage(tag: String): Script = when (base(tag)) {
        "ar", "fa", "ur", "ps", "ckb", "sd", "ug" -> Script.ARABIC
        "ru", "uk", "be", "bg", "sr", "mk", "kk", "ky", "mn", "tg" -> Script.CYRILLIC
        "hi", "mr", "ne", "sa" -> Script.DEVANAGARI
        "zh", "yue" -> Script.HAN
        "ja" -> Script.KANA
        "ko" -> Script.HANGUL
        "he", "iw", "yi" -> Script.HEBREW
        "el" -> Script.GREEK
        "th" -> Script.THAI
        else -> Script.LATIN
    }

    /** The language written in [script] among [known] (keyboards, then the phone's languages), else the script's most spoken one. */
    fun languageFor(script: Script, known: List<String>): String =
        known.mapNotNull(::tag).firstOrNull { scriptOfLanguage(it) == script || (script == Script.HAN && base(it) == "ja") }
            ?: when (script) {
                Script.ARABIC -> "ar"
                Script.LATIN -> "en"
                Script.CYRILLIC -> "ru"
                Script.DEVANAGARI -> "hi"
                Script.HAN -> "zh"
                Script.KANA -> "ja"
                Script.HANGUL -> "ko"
                Script.HEBREW -> "he"
                Script.GREEK -> "el"
                Script.THAI -> "th"
            }

    /**
     * The languages to try, best first: the person's choice; with Auto, the keyboard in use, then
     * the language the conversation is written in, then the phone's languages, then the app's.
     */
    fun candidates(
        choice: String,
        keyboard: String?,
        recent: List<String>,
        keyboards: List<String>,
        preferred: List<String>,
        app: String,
    ): List<String> {
        val list = mutableListOf<String>()
        if (choice != AUTO) tag(choice)?.let(list::add)
        tag(keyboard)?.let(list::add)
        script(recent)?.let { list += languageFor(it, keyboards + preferred) }
        list += preferred.mapNotNull(::tag)
        tag(app)?.let(list::add)
        val seen = mutableSetOf<String>()
        return list.filter { seen.add(it.lowercase()) }
    }

    /** The region a bare language is asked for in, so the recognizer gets a full locale. */
    private val usualRegion = mapOf(
        "ar" to "SA", "en" to "US", "es" to "ES", "fr" to "FR", "de" to "DE", "pt" to "BR", "zh" to "CN",
        "it" to "IT", "nl" to "NL", "ru" to "RU", "ja" to "JP", "ko" to "KR", "hi" to "IN", "tr" to "TR",
    )

    /** A tag with a region: `ar` → `ar-SA`; one that has a region, or a language without a usual one, as it is. */
    fun withRegion(tag: String): String =
        if ('-' in tag) tag else usualRegion[base(tag)]?.let { "$tag-$it" } ?: tag

    /** What the recognizer may detect and switch between with Auto: the keyboards' languages and the phone's. */
    fun allowed(first: List<String>, keyboards: List<String>, preferred: List<String>): List<String> {
        val seen = mutableSetOf<String>()
        return (first + keyboards + preferred).mapNotNull(::tag).map(::withRegion).filter { seen.add(base(it)) }.take(10)
    }

    /**
     * What the microphone's menu offers: the keyboards' languages, then the popular ones, each
     * once, and only languages the phone recognizes when it says ([supported] null: all).
     */
    fun menu(keyboards: List<String>, supported: List<String>?): List<String> {
        val bases = supported?.map(::base)?.toSet()
        val seen = mutableSetOf<String>()
        return (keyboards.mapNotNull(::tag) + popular).filter { language ->
            seen.add(base(language)) && (bases == null || base(language) in bases)
        }
    }

    /** The tiny mark on the microphone while a language is chosen: `AR`, `EN`, `ZH`. */
    fun badge(choice: String): String? = if (choice == AUTO) null else base(choice).uppercase(Locale.ROOT)

    /** A language's name in the reader's language: «العربية», "Arabic". */
    fun name(tag: String, reader: String): String {
        val readerLocale = Locale.forLanguageTag(reader)
        val name = Locale.forLanguageTag(tag).getDisplayName(readerLocale).ifBlank { tag }
        return name.replaceFirstChar { it.titlecase(readerLocale) }
    }

    /** A stored choice as it reads now: up to 1.1.x the choices were APP (the app's language — what made an English app hear Arabic as English), AR and EN. */
    fun stored(value: String?): String = when (value) {
        null, "", "APP", AUTO -> AUTO
        "AR" -> "ar"
        "EN" -> "en"
        else -> tag(value) ?: AUTO
    }

    /** The phone's voice for a reply: the language its letters are written in, as the phone knows it; English when it cannot tell. */
    fun replyLanguage(text: String, known: List<String>): String =
        script(listOf(text))?.let { base(languageFor(it, known)) } ?: "en"
}
